feat: initial `Operator::from_uri` implementation

jorgehermo9 opened the pull request:
Relates to #5445

Left some doubts as `// TODO` in the source. I have little experience contributing to this repo so I'm sorry if there are _a lot_ of doubts about this. Just want to be sure all the changes of this PR aligns with the current codebase. Please take a look to all the `TODOs` I left when reviewing.

I would like to add more tests, but I don't know in which place those should be placed. The `core/tests` folder seems like a good place, but I don't find any place suitable, as placing those in `core/tests/behaviour` seems weird to me. But as this implies various components, maybe we can have a `core/tests/integration`? Although I would like to write some unit tests at `core/src/types/builder.rs`, `core/src/types/operator/builder.rs` and `core/src/types/operator/registry.rs`, but didn't any existing unit tests there.

In this PR I implemented a single `Configurator::from_uri` method, which will serve as default and takes only the query parameters as options. Services which need a more specific configuration such as s3 or azblob can be implemented in follow-up PRs.

I also have pending documentating all the newly added public API, but will do that after an initial review round.

Thank you very much.

jorgehermo9 committed:
feat: initial OperatorRegistry implementation

jorgehermo9 committed:
feat: continue with operator registry implementation

jorgehermo9 committed:
feat: register enabled services & set global registry

jorgehermo9 committed:
feat: glue together Operator and global OperatorRegistry

jorgehermo9 committed:
feat: glue together Operator and global OperatorRegistry

jorgehermo9 committed:
feat: implement Configurator::from_uri

jorgehermo9 committed:
test: add small doctest to Operator::from_uri

jorgehermo9 committed:
chore: add comment

jorgehermo9 committed:
chore: remove changes

jorgehermo9 commented on core/src/types/operator/builder.rs:
````diff
@@ -95,6 +97,22 @@ impl Operator {
+    ///
+    /// Ok(())
+    /// ```
+    pub fn from_uri(
````
should this method be called `via_uri`? 

The `from_iter` method returns an `OperatorBuilder` instead of `Operator` https://github.com/apache/opendal/blob/6eed396b7e48abf42b7548f719ee987840df1bf4/core/src/types/operator/builder.rs#L147

and the `via_iter` method returns an `Operator` (as this `from_uri` method does) https://github.com/apache/opendal/blob/6eed396b7e48abf42b7548f719ee987840df1bf4/core/src/types/operator/builder.rs#L185

Should we include two new methods? `via_uri` and `from_uri`?

Xuanwo replied:
`from_uri` is good enough, all `via_xxx` API may be removed in the future.

jorgehermo9 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+fn operator_factory_from_configurator<C: Configurator>() -> OperatorFactory {
```
included this private function to reduce boilerplate of operator factory generation. Any suggestion about it or is it ok to have it here?

Xuanwo replied:
I prefer to remove `operator_factory_from_configurator`.

jorgehermo9 replied:
implemented it as a local closure in [3b80f70](https://github.com/apache/opendal/pull/5482/commits/3b80f703c9dab2b8db32cd81cf3e5071af26c35b)

jorgehermo9 committed:
chore: remove changes

jorgehermo9 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,226 @@
+use std::cell::LazyCell;
```
I added this file inside the `crate::types::operator::registry` module. Is it ok? I thought about adding a `crate::types::operator_registry`, but it seemed better this way.

jorgehermo9 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,226 @@
+
+// TODO: thread local or use LazyLock instead? this way the access is lock-free
+// TODO: should we expose the `GLOBAL_OPERATOR_REGISTRY` as public API at `crate::types::operator::GLOBAL_OPERATOR_REGISTRY`?
+thread_local! {
```
What is the preferred way of having a global static variable such as this?

I prefer to have it `thread_local` so there is not need for a `LazyLock`, we can use `LazyCell` instead (`LazyCell` is lock-free but `LazyLock` is not, it synchronizes access through threads)

asukaminato0721 replied:
just curious, why do we need static var here?

jorgehermo9 replied:
In order to have a global initialized registry and to not have to initialize it every time whenever we want to parse a uri

https://github.com/apache/opendal/blob/088d7d029b233f4c0b17962277db4a9490a55da3/core/src/types/operator/builder.rs#L118

jorgehermo9 committed:
chore: add license header

jorgehermo9 committed:
chore: fix typo

jorgehermo9 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,242 @@
+// TODO: thread local or use LazyLock instead? this way the access is lock-free
+// TODO: should we expose the `GLOBAL_OPERATOR_REGISTRY` as public API at `crate::types::operator::GLOBAL_OPERATOR_REGISTRY`?
+thread_local! {
+    pub static GLOBAL_OPERATOR_REGISTRY: LazyCell<OperatorRegistry> = LazyCell::new(|| OperatorRegistry::with_enabled_services());
```
MSRV check fails due to the usage of `LazyCell`. Should we update the MSRV or use another thing instead? 

I see this `TODO` about the `once_cell` crate usage: https://github.com/apache/opendal/blob/52c96bb8e8cb4d024ccab1f415c4756447c726dd/core/Cargo.toml#L248

If the replacement is planned, I think it would be better to use `LazyCell` than `once_cell::Lazy` in new code like this one, to not increase technical debt.

<img width="640" alt="image" src="https://github.com/user-attachments/assets/c90bdc76-f98c-4f96-a437-a0b48e09903e" />

jorgehermo9 committed:
fix: clippy lint

jorgehermo9 committed:
retrigger ci

Xuanwo commented on core/Cargo.toml:
```diff
@@ -255,6 +255,9 @@ reqwest = { version = "0.12.2", features = [
 tokio = { version = "1.27", features = ["sync", "io-util"] }
+# TODO: I added this dependency in order to not re-implement the Url::query_pairs function.
+# Is it okay to include it? (we are using that crate in the ofs binary https://github.com/apache/opendal/blob/main/bin/ofs/Cargo.toml#L45
+url = "2.5.4"
```
Please use `http::uri` instead. During the development of `core`, there is no need to reference other download stream projects. Simply make the best decision based on the current `core` code.

jorgehermo9 replied:
Ok, no problem. Going to implement some utility methods in `crate::raw::http_util` module to replace this usage.

Xuanwo commented on core/src/types/builder.rs:
```diff
@@ -137,6 +137,51 @@ pub trait Configurator: Serialize + DeserializeOwned + Debug + 'static {
         })
     }
 
+    // TODO: should we split `from_uri` into two functions? `from_uri` and `from_uri_opts`?
```
Doesn't worth it.

Xuanwo commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,242 @@
+// Register only services in `Scheme::enabled()`
+
+pub struct OperatorRegistry {
+    // TODO: add Arc<Mutex<...>> to make it cheap to clone + thread safe? or is it not needed?
```
It makes sense to me; we rarely change it or keep the lock engaged for long.

jorgehermo9 replied:
done in b6dd1dc1c354c8b7c843bfebd358780626ee0d46

Xuanwo commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,242 @@
+}
+
+impl OperatorRegistry {
+    pub fn new() -> Self {
```
Let's initialize the registry here directly.

jorgehermo9 replied:
Ok, done in e887d246

Derived a `Default` implementation for `OperatorRegistry` that rely in `HashMap::default` so users can get an empty `OperatorRegistry` if they need it.

Or should the `Default` implementation include all the registered services as the `OperatorRegistry::new`? If so, should we include a `OpereatorRegistry::empty` so users can get an empty `OperatorRegistry`?

Xuanwo commented on core/src/types/operator/registry.rs:
````diff
@@ -0,0 +1,242 @@
+        // ```
+        // and the apply_for_all_services macro would gate every statement behind the corresponding feature gate
+        // This seems to not be the place where we should have a "list of enabled services".
+        #[cfg(feature = "services-aliyun-drive")]
````
I believe every service should have its own registration functions. There are two reasons:

- The service may have different schemes for registration. For example, S3 registers as `s3`, `s3a`, `minio`, `r2`, and so on, while GCS registers as `gcs`, `gs`, and so forth.
- Provide a register that simplifies integration with OpenDAL for external users.

jorgehermo9 replied:
Ok, going to move the registration to each service `Builder`.

something like
```rust
impl Builder for S3Builder{

  pub fn register_in_registry(registry: &mut OperatorRegistry){
   registry.register("s3",s3_builder)
   registry.register("r2", r2_builder)
  }
}
```

I wonder if `s3` and `r2` (or `s3a`, `minio`, etc) have different ways of building, if so, we can use different `OperatorFactory` such as `s3_builder`, `r2_builder`, etc.

jorgehermo9 replied:
Implemented that in 98f8022f. Do you prefer it that way?

Please note that I used a default `Builder::register_in_registry` for all services. 

> For example, S3 registers as s3, s3a, minio, r2, and so on, while GCS registers as gcs, gs, and so forth.
This can be addressed in follow-up PR's (as discussed [here](https://github.com/apache/opendal/issues/5445#issuecomment-2564991993)) and then `S3Builder` should override the default `Builder::register_in_registry` implementation to register the custom schemes/factories if needed.

jorgehermo9 committed:
feat: drop usage of the url crate

jorgehermo9 committed:
chore: remove TODO

jorgehermo9 committed:
feat: wrap OperatorRegistry in Arc Mutex

jorgehermo9 committed:
feat: Initialize operator registry in new method

jorgehermo9 committed:
chore: remove todo

jorgehermo9 committed:
Merge branch 'main' into feature/operator-from-uri

jorgehermo9 committed:
feat: move service registration to builder methods

jorgehermo9 commented on core/src/types/builder.rs:
```diff
@@ -56,6 +58,15 @@ pub trait Builder: Default + 'static {
+
+    /// Register this builder in the given registry.
+    fn register_in_registry(registry: &mut OperatorRegistry) {
+        let operator_factory: OperatorFactory = |uri, options| {
```
Had to add the type annotation for `operator_factory` because the compiler could not infer it correctly

```console
error[E0308]: mismatched types
  --> src/types/builder.rs:68:55
   |
68 |         registry.register(Self::SCHEME.into_static(), operator_factory)
   |                                                       ^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected fn pointer `for<'a> fn(&'a _, HashMap<_, _>) -> std::result::Result<_, _>`
              found fn pointer `fn(&_, HashMap<_, _>) -> std::result::Result<_, _>`

For more information about this error, try `rustc --explain E0308`.
warning: `opendal` (lib) generated 2 warnings
error: could not compile `opendal` (lib) due to 1 previous error; 2 warnings emitted
```

jorgehermo9 commented on core/src/types/builder.rs:
```diff
@@ -56,6 +58,15 @@ pub trait Builder: Default + 'static {
+            let builder = Self::Config::from_uri(uri, options)?.into_builder();
+            Ok(Operator::new(builder)?.finish())
+        };
+        registry.register(Self::SCHEME.into_static(), operator_factory)
```
As a default implementation, we simply register the associated `SCHEME` with this builder. Is this ok?

asukaminato0721 replied:
I guess we can add tests, so if test works, then it works.

jorgehermo9 commented on core/src/types/builder.rs:
```diff
@@ -56,6 +58,15 @@ pub trait Builder: Default + 'static {
     /// Consume the accessor builder to build a service.
     fn build(self) -> Result<impl Access>;
+
+    /// Register this builder in the given registry.
```
Is this doc-comment adequate?

jorgehermo9 committed:
chore: fix comment typo

jorgehermo9 committed:
chore: fix comment typo

jorgehermo9 commented:
hi @Xuanwo, thank you very much for your review. I answered all the comments and I'm ready for another round.

jorgehermo9 committed:
chore: add comment

jorgehermo9 commented:
Don't worry, take your time! Just pinging to notify, but no hurries! Thank you very much for your hard work

jorgehermo9 commented:
Hi @Xuanwo, just a little remind of this in case this got forgotten. I have more time right now to keep working on this

asukaminato0721 committed:
Merge branch 'main' into feature/operator-from-uri

asukaminato0721 committed:
fix bracket

jorgehermo9 commented:
Hi @asukaminato0721  are you interested on this PR? I was waiting for @Xuanwo next review, I'm open to keep working on this anytime

asukaminato0721 commented:
https://github.com/apache/opendal/pull/5616 so libsql can be removed.

asukaminato0721 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,227 @@
+        Ipmfs::register_in_registry(&mut registry);
+        #[cfg(feature = "services-icloud")]
+        Icloud::register_in_registry(&mut registry);
+        #[cfg(feature = "services-libsql")]
+        Libsql::register_in_registry(&mut registry);
```
```suggestion
```

asukaminato0721 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,227 @@
+        Sled::register_in_registry(&mut registry);
+        #[cfg(feature = "services-sqlite")]
+        Sqlite::register_in_registry(&mut registry);
+        #[cfg(feature = "services-supabase")]
+        Supabase::register_in_registry(&mut registry);
```
```suggestion
```

asukaminato0721 commented on core/src/types/operator/mod.rs:
```diff
@@ -32,3 +32,11 @@ pub use info::OperatorInfo;
+// TODO: warning as not used. How can we expose them as public api?
+pub use registry::OperatorFactory;
+pub use registry::OperatorRegistry;
+pub use registry::GLOBAL_OPERATOR_REGISTRY;
```
```suggestion
```

remove it for now so clippy will happy.

asukaminato0721 committed:
clippy

some temp fix

asukaminato0721 commented on core/src/raw/http_util/uri.rs:
```diff
@@ -64,6 +64,25 @@ pub fn percent_decode_path(path: &str) -> String {
     }
 }
 
+/// query_pairs will parse a query string encoded as key-value pairs separated by `&` to a vector of key-value pairs.
+/// It also does percent decoding for both key and value.
+///
+/// Note that `?` is not allowed in the query string, and it will be treated as a part of the first key if included.
+pub fn query_pairs(query: &str) -> Vec<(String, String)> {
+    query
+        .split('&')
+        .filter_map(|pair| {
+            let mut iter = pair.splitn(2, '=');
+            // TODO: should we silently ignore invalid pairs and filter them out without the user noticing?
+            // or should we return an error in the whole `query_pairs` function so the caller knows it failed?
+            let key = iter.next()?;
+            let value = iter.next().unwrap_or("");
+            Some((key, value))
+        })
+        .map(|(key, value)| (percent_decode_path(key), percent_decode_path(value)))
+        .collect()
+}
+
```
we have

```rs
pub struct QueryPairsWriter {
```

now, use this.

jorgehermo9 replied:
I don't get what you mean. `QueryPairsWriter` is used to *write query param pairs* to an url, not for reading them. The function `query_pairs` this PR adds is to *read query param pairs` from an url

asukaminato0721 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    pub fn register(&mut self, scheme: &str, factory: OperatorFactory) {
+        // TODO: should we receive a `&str` or a `String`? we are cloning it anyway
```
if you need to take the ownship, then use String

ref https://users.rust-lang.org/t/idiomatic-string-parmeter-types-str-vs-asref-str-vs-into-string/7934/4

jorgehermo9 replied:
Using `Into<String>` so the caller has less boilerplate. Thanks for the post

asukaminato0721 commented on core/src/types/operator/mod.rs:
```diff
@@ -32,3 +32,10 @@ pub use info::OperatorInfo;
+// TODO: should we make the registry module public or export the OperatorFactory and OperatorRegistry
+// types directly?
+mod registry;
+// TODO: warning as not used. How can we expose them as public api?
```
if user will use it: pub
if only in crate: pub(crate)
if not: not pub

asukaminato0721 commented on core/src/types/builder.rs:
```diff
@@ -137,6 +148,32 @@ pub trait Configurator: Serialize + DeserializeOwned + Debug + 'static {
+        // this is very similar to https://github.com/apache/opendal/blob/52c96bb8e8cb4d024ccab1f415c4756447c726dd/bin/ofs/README.md?plain=1#L45
+        let query_pairs = parsed_uri.query().map(query_pairs).unwrap_or_default();
+
+        // TODO: should we log a warning if the query_pairs vector is empty?
```
if it wont cause problem, then ignore it, or `Err`.

asukaminato0721 commented on core/src/types/builder.rs:
```diff
@@ -137,6 +148,32 @@ pub trait Configurator: Serialize + DeserializeOwned + Debug + 'static {
+                .set_source(err)
+        })?;
+
+        // TODO: I have some doubts about this default implementation
```
consider raise an issue if there is a bug.

asukaminato0721 commented on core/src/types/builder.rs:
```diff
@@ -137,6 +148,32 @@ pub trait Configurator: Serialize + DeserializeOwned + Debug + 'static {
+        // TODO: we are not interpreting the host or path params
+        // the `let op = Operator::from_uri("fs:///tmp/test", vec![])?;` statement from the RFC wont work.
+        // instead we should use `let op = Operator::from_uri("fs://?root=/tmp/test", vec![])?;` as done
+        // in `ofs`. The `fs` service should override this default implementation if it wants to use the host or path params?
```
https://github.com/apache/opendal/pull/5482/files#r1899962804

asukaminato0721 commented on core/src/types/operator/builder.rs:
```diff
@@ -21,6 +21,8 @@ use std::sync::Arc;
 use crate::layers::*;
 use crate::raw::*;
 use crate::*;
+// TODO: is this import path idiomatic to the project?
```
see how other code do.

asukaminato0721 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,223 @@
+// In order to reduce boilerplate, we should return in this function a `Builder` instead of operator?.
+pub type OperatorFactory = fn(&str, HashMap<String, String>) -> Result<Operator>;
+
+// TODO: the default implementation should return an empty registry? Or it should return the initialized
```
or we can remove the `Default` ? since idk how `Default` means here.

jorgehermo9 replied:
Yes, had the same doubts. I will change the api a little bit

asukaminato0721 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,223 @@
+        uri: &str,
+        options: impl IntoIterator<Item = (String, String)>,
+    ) -> Result<Operator> {
+        // TODO: we use the `url::Url` struct instead of `http:Uri`, because
```
this is not a todo imo

asukaminato0721 commented:
so many todo...

keep the code change small, clippy happy, test pass, and the code style unified (you can use llm for this kind of idea)

jorgehermo9 commented:
Thank you very much for your review @asukaminato0721! I usually leave the TODOs in code as comments to the reviewer, will address everything asap!

jorgehermo9 committed:
test: add tests for query_pairs

jorgehermo9 committed:
feat: address comments & todos

jorgehermo9 commented on core/src/types/builder.rs:
```diff
@@ -137,6 +148,15 @@ pub trait Configurator: Serialize + DeserializeOwned + Debug + 'static {
+    fn from_uri(uri: &Uri, options: impl IntoIterator<Item = (String, String)>) -> Result<Self> {
+        let query_pairs = uri.query().map(query_pairs).unwrap_or_default();
+
+        let merged_options = query_pairs.into_iter().chain(options);
```
Unsure about 
`let merged_options = query_pairs.into_iter().chain(options);`
or
`let merged_options = options..chain(query_pairs.into_iter());

Which one should take precedence? query_pairs should overwrite the options, or should the options overwrite the query_pairs? I'm thinking on the case that query_pairs and options contain the same key  

Right now, I think that the behaviour with `query_pairs.into_iter().chain(options)` is that `options` takes precedence and them overwrite whatever goes in `query_pairs` if a key is present in both

jorgehermo9 committed:
feat: fast committed

jorgehermo9 commented on core/src/types/operator/registry.rs:
```diff
@@ -0,0 +1,204 @@
+use crate::*;
+
+/// TODO: document this
+pub static GLOBAL_OPERATOR_REGISTRY: LazyLock<OperatorRegistry> =
```
should we expose this as public api @xuanwo? or should we `pub(crate)` instead

jorgehermo9 commented on core/src/types/operator/mod.rs:
```diff
@@ -32,3 +32,8 @@ pub use info::OperatorInfo;
 pub mod operator_futures;
+
+mod registry;
+pub use registry::OperatorFactory;
+pub use registry::OperatorRegistry;
+pub use registry::GLOBAL_OPERATOR_REGISTRY;
```
should we expose these 3 as public api @Xuanwo? or should we pub(crate) them instead?

Or should we only expose `OperatorFactory` and `OperatorRegistry` but not `GLOBAL_OPERATOR_REGISTRY`?

jorgehermo9 committed:
feat: derive default

jorgehermo9 commented:
Hi @Xuanwo, I worked a bit more on this and I think it is ready for another review round. Once the implementation looks good to you, I will work on tests and documentation, is this okay?

@asukaminato0721 I think I addressed most of your comments

jorgehermo9 commented on core/src/types/builder.rs:
```diff
@@ -137,6 +148,15 @@ pub trait Configurator: Serialize + DeserializeOwned + Debug + 'static {
     }
 
+    /// TODO: document this.
+    fn from_uri(uri: &Uri, options: impl IntoIterator<Item = (String, String)>) -> Result<Self> {
```
Is it okay to accept `&Uri` here? This way, we don't have to parse the uri again (it is only done [here](https://github.com/apache/opendal/blob/5ecb178ec7465cfd69f7b36d79429e588233c55a/core/src/types/operator/registry.rs#L181) right now)

Or should we change it to `&str` and parse the uri twice?

jorgehermo9 committed:
feat: simplify operator from config

jorgehermo9 committed:
feat: fast committed

jorgehermo9 committed:
feat: fast committed

jorgehermo9 committed:
Merge branch 'main' into feature/operator-from-uri

DerGut commented on core/src/raw/http_util/uri.rs:
```diff
@@ -64,6 +64,18 @@ pub fn percent_decode_path(path: &str) -> String {
+/// It also does percent decoding for both key and value.
+///
+/// Note that `?` is not allowed in the query string, and it will be treated as a part of the first key if included.
+pub fn query_pairs(query: &str) -> Vec<(String, String)> {
```
Does this need to be exposed as a public API?

IIUC the only public entrypoint to the registry is `OperatorRegistry::parse()` which takes a URI as a string. Is there a need for users to manually parse a URI string and split its query into parts for further customization? If not, I'd suggest to keep `query_pairs()` `pub(crate)` (at least for now) to keep the API surface minimal.

Query parsing isn't really the main responsibility of OpenDAL so IMO it would make sense to keep such utilities internal if we can.

jorgehermo9 replied:
You are absolutely right! Thanks for noticing this. Will change it asap

jorgehermo9 committed:
feat: make query_pairs pub(crate)

jorgehermo9 committed:
feat: make query_pairs pub(crate)

Xuanwo commented:
Sorry, @jorgehermo9, for keeping you waiting so long. I plan to start working on this soon.

The main issue preventing us from moving forward smoothly is that we haven’t yet decided how to split the crates and design our registry.

I think we can break this PR into two parts: one for splitting opendal, which will divide opendal into `opendal`, `opendal-core`, `opendal-services-xxx`, and `opendal-layers-xxx` as we discussed in https://github.com/apache/opendal/discussions/5206. In this phase, we’ll figure out how to split opendal and design a registry mechanism that is easy to use and works well for opendal after the split. Based on that, we can design a `from_uri` API that can reuse our existing framework. At that point, adding the `from_uri` API should be simple and won’t introduce any breaking changes.

I led you down the wrong path because I used to think it was fine to add `from_uri` first and then perform the splits. However, during the review of this PR, I finally realized that I had misunderstood these two processes. In fact, we can't have a proper `from_uri` unless we know how the opendal split and opendal registry will look.

jorgehermo9 committed:
Merge branch 'main' into feature/operator-from-uri

Xuanwo commented:
Thank you for working on this! Now from_uri has been implemented, let's close.
